middle_end/flambda/flambda_to_clambda.ml

```diff
@@ -321,130 +335,207 @@ let rec to_clambda t env (flam : Flambda.t) : Clambda.ulambda =
         match body with
         | Uvar _ | Uconst _ -> true
         | _ -> false
       in
-      if is_trivial then body
-      else Uregion body
+      if is_trivial then body, body_layout
+      else Uregion body, body_layout
   | Tail body ->
-      let body = to_clambda t env body in
+      let body, body_layout = to_clambda t env body in
       let is_trivial =
         match body with
         | Uvar _ | Uconst _ -> true
         | _ -> false
       in
-      if is_trivial then body
-      else Utail body
-  | Proved_unreachable -> Uunreachable
+      if is_trivial then body, body_layout
+      else Utail body, body_layout
+  | Proved_unreachable -> Uunreachable, Lambda.layout_bottom
 
-and to_clambda_named t env var (named : Flambda.named) : Clambda.ulambda =
+and to_clambda_named t env var (named : Flambda.named) : Clambda.ulambda * Lambda.layout =
   match named with
-  | Symbol sym -> to_clambda_symbol env sym
-  | Const (Int n) -> Uconst (Uconst_int n)
-  | Const (Char c) -> Uconst (Uconst_int (Char.code c))
+  | Symbol sym -> to_clambda_symbol env sym, Lambda.layout_any_value
+  | Const (Int n) -> Uconst (Uconst_int n), Lambda.layout_int
+  | Const (Char c) -> Uconst (Uconst_int (Char.code c)), Lambda.layout_int
   | Allocated_const _ ->
     Misc.fatal_errorf "[Allocated_const] should have been lifted to a \
         [Let_symbol] construction before [Flambda_to_clambda]: %a = %a"
       Variable.print var
       Flambda.print_named named
   | Read_mutable mut_var ->
-    begin try Uvar (Env.ident_for_mutable_var_exn env mut_var)
+    begin try
+      let mut_var, layout = Env.ident_for_mutable_var_exn env mut_var in
+      Uvar mut_var, layout
     with Not_found ->
       Misc.fatal_errorf "Unbound mutable variable %a in [Read_mutable]: %a"
         Mutable_variable.print mut_var
         Flambda.print_named named
     end
   | Read_symbol_field (symbol, field) ->
-    Uprim (Pfield field, [to_clambda_symbol env symbol], Debuginfo.none)
+    Uprim (Pfield field, [to_clambda_symbol env symbol], Debuginfo.none),
+    Lambda.layout_any_value
   | Set_of_closures set_of_closures ->
-    to_clambda_set_of_closures t env set_of_closures
+    to_clambda_set_of_closures t env set_of_closures,
+    Lambda.layout_any_value
   | Project_closure { set_of_closures; closure_id } ->
     (* Note that we must use [build_uoffset] to ensure that we do not generate
        a [Uoffset] construction in the event that the offset is zero, otherwise
        we might break pattern matches in Cmmgen (in particular for the
        compilation of "let rec"). *)
+    let set_of_closures_expr, _layout_set_of_closures =
+      subst_var env set_of_closures
+    in
     check_closure t (
       build_uoffset
-        (check_closure t (subst_var env set_of_closures)
+        (check_closure t set_of_closures_expr
            (Flambda.Expr (Var set_of_closures)))
         (get_fun_offset t closure_id))
-      named
+      named,
+    Lambda.layout_any_value
```

Comment:
I'd use `layout_function` (soon to be replaced by `layout_closure` once everything is merged) here, since it returns a closure.

Reply:
Fixed